README
LFS Book README

This document is meant to instruct the user on how to convert the book's XML 
source to other formats (e.g. HTML, PDF, PS and TXT).  First, if you have not 
already done so, please read the INSTALL file for instructions on how to install
the required software.  If you have already completed the steps outlined in the
INSTALL file, then continue reading for examples how to convert these files into
various other formats.

In all examples, setting the parameter REV=systemd is needed to build the 
systemd version of the book.

XML to XHTML:
-------------
make BASEDIR=/path/to/output/location


XML to single file XHTML (nochunks):
------------------------------------
make BASEDIR=/path/to/output/location nochunks

XML to TXT
----------
Follow the instructions for nochunks and then run:
lynx -dump /path/to/nochunks >/path/to/output

XML to PDF:
-----------
make BASEDIR=/path/to/output/location pdf




udev-lfs/ChangeLog
2017-10-27  <bdubbs@linuxfromscratch.org>
   * 83-cdrom-symlinks.rules: Adjust test for writable root directory.

2014-03-02  <bdubbs@linuxfromscratch.org>
   * Makefile.lfs: Remove build procedures leaving only LFS rules

2014-02-16  <bdubbs@linuxfromscratch.org>
   * Makefile.lfs: Adjust warning flags.  Adjust linking rules to
     allow build with util-linux in LFS's /tools directory.

2010-01-26	Matt Burgess <matthew@linuxfromscratch.org>
	* 55-lfs.rules: Remove lots of rules that have been merged upstream.
	* 61-cdrom.rules: Remove as upstream has a replacement.
	* doc/*: Cleanup and rewrite now that the rules are much simpler.

2009-09-25	Bryan Kadzban <bryan@linuxfromscratch.org>
	* 55-lfs.rules: Make the RTC rule (which runs setclock) work for people
	  that don't use the RTC-class driver -- add another copy of the rule,
	  using KERNEL=="rtc" instead of SUBSYSTEM=="rtc".  Also move the ACTION
	  match before the assignment to MODE.

2009-05-23	Bruce Dubbs <bdubbs@linuxfromscratch.org>
	* 55-lfs.rules: Set the rtc by udev upon boot.  Removed aio device from
	  udev configuration so defaults are used.

2009-05-16	Bryan Kadzban <bryan@linuxfromscratch.org>
	* 55-lfs.rules: Adopt udev permissions for random, urandom, kmsg, and
	  input devices (including psaux, which probably doesn't exist anymore
	  anyway).

2009-05-16	Matt Burgess <matthew@linuxfromscratch.org>
	* 55-lfs.rules: Remove the block rule as it was only setting
	  permissions that Udev will set by default
	* 55-lfs.rules: Remove the rfcomm rule as the group it sets is already
	  included in upstream's rule
	* 55-lfs-rules: Remove the tape rules as the group they sets is already
	  included in upstream's rules

2009-03-15	Bryan Kadzban <bryan@linuxfromscratch.org>
	* 55-lfs.rules: Remove the ISDN-handling rules (replaced with upstream's
	  version, which has a better match via SUBSYSTEM).
	* 55-lfs.rules: Replace "uucp" with "dialout" now that udev has decided
	  to standardize on a Debian-like setup (uucp is for UUCP daemons to use
	  for privilege separation; dialout is for users and devices).

2008-12-07  Bruce Dubbs <bdubbs@linuxfromscratch.org>
	* 55-lfs.rules: Change one line to use continuation so it does
	  not overflow the book's width.

2008-12-07  Bryan Kadzban  <bryan@linuxfromscratch.org>
	* 55-lfs.rules: Remove several rules that are either provided by upstream,
	  or that don't have any effect (there is no /dev/js or /dev/djs according
	  to devices.txt).
	* 55-lfs.rules, 61-cdrom.rules: Replace ACTION=="add" with "add|change"
	  everywhere, per upstream's general request.

2008-11-11  Bryan Kadzban  <bryan@linuxfromscratch.org>
	* 55-lfs.rules, 61-cdrom.rules: Fix comments to use the right rule
	  numbers.  Thanks to Trent Shea for the fix.  Closes #2278.

2008-10-15  DJ Lucas <dj@linuxfromscratch.org>
	* 55-lfs.rules: Override default perms on floppy disk devices provided
	  by 50-udev-default.rules.  Thanks to Bruce Dubbs for the fix.  Closes
	  LFS ticket #2076. 

2008-05-21  Bryan Kadzban  <bryan@linuxfromscratch.org>
	* 81-firmware.rules, doc/81-firmware.txt: Remove.  This rule is
	  already handled by udev's 50-udev-default.rules file.  The docs
	  can be added back later if needed.
	* Makefile: Don't install the above deleted files.
	* 55-lfs.rules: Remove the device-mapper rule, since the upstream
	  50-udev-default.rules file handles it properly.

2008-04-02  Bryan Kadzban  <bryan@linuxfromscratch.org>
	* Makefile: Make the package DESTDIR-friendly by installing the docs
	  into a fixed $(PREFIX)/share/doc/udev-config directory, instead of
	  trying to figure out what version of udev was just installed.

2007-10-30  Bryan Kadzban  <bryan@linuxfromscratch.org>
	* 55-lfs.rules: Since the dialout group was renamed to uucp, delete
	  the rules that override upstream's assignment of the dialout group.
	  Replace "dialout" with "uucp" on the remaining rules.

2007-10-29  Bryan Kadzban  <bryan@linuxfromscratch.org>
	* 51-lfs.rules: Move to 55-lfs.rules.
	* doc/51-lfs.rules: Move to doc/55-lfs.rules.

2007-10-27  Bryan Kadzban  <bryan@linuxfromscratch.org>
	* 25-lfs.rules: Some rules in here are duplicates of rules from udev's
	  new 50-udev-defaults.rules file; remove them. Override permissions
	  where needed elsewhere (and document overrides in comments).
	* 26-modprobe.rules: Provided by udev's 80-drivers.rules and built-in
	  modaliases for SCSI device-type modules in 2.6.22+ kernels; remove.

	* 25-lfs.rules: Move to 51-lfs.rules.
	* 27-firmware.rules: Move to 81-firmware.rules.
	* 81-cdrom.rules: Move to 61-cdrom.rules.

	* doc/25-lfs.txt: Rename to 51-lfs.txt.
	* doc/26-modprobe.txt: Rename to 80-drivers.txt, and modify to reflect
	  the upstream rules.
	* doc/27-firmware.txt: Rename to 81-firmware.txt.
	* doc/81-cdrom.txt: Rename to 61-cdrom.txt.

	* Makefile: Use new filenames.

2007-07-31  Dan Nicholson <dnicholson@linuxfromscratch.org>
	* 25-lfs.rules: Changed the usb_device rule tto create /dev/bus/usb
	  nodes if the usb_device in 2.6.22+ kernels. The rule was also changed
	  to only trigger on "add" events. This change is not backwards
	  compatible with older kernels.

2007-06-12  Dan Nicholson <dnicholson@linuxfromscratch.org>
	* 25-lfs.rules: Fix the CPUID nodes from cpu/%n/cpu to cpu/%n/cpuid,
	  which is what's expected in userspace apps such as x86info.

2007-06-08  Bryan Kadzban  <bryan@linuxfromscratch.org>
	* 25-lfs.rules: Add rules for DVB devices (create nodes in /dev/dvb/)
	  and floppies (create extra nodes based on CMOS type), copied from
	  the SuSE rules file.  Thanks to Alexander Patrakov for the bugreport.

2007-03-04  Matthew Burgess	<matthew@linuxfromscratch.org>
	* Makefile: Use `udevd --version' to work out what version of Udev is
	  installed (requires Udev >= 106)

2007-01-02  Bryan Kadzban  <bryan@linuxfromscratch.org>
	* 25-lfs.rules: Change CPU devices (cpu, msr, microcode) to be in
	  /dev/cpu/ and /dev/cpu/N/, to match Documentation/devices.txt.


process-scripts.sh
#!/bin/bash

# Boot scripts
for s in bootscripts/lfs/init.d/*       \
         bootscripts/lfs/sysconfig/*    \
         bootscripts/lfs/lib/services/* \
         bootscripts/lfs/sbin/*         \
         udev-lfs/*.rules
do
  script=$(basename $s)
  
  # Skip directories
  [ $script == 'network-devices' ] && continue
  #[ $script == 'services'        ] && continue

  # Disambiguate duplicate file names
  [ $s == 'bootscripts/lfs/sysconfig/rc'         ] && script='rc-sysinit'; 
  [ $s == 'bootscripts/lfs/sysconfig/modules'    ] && script='modules-sysinit'; 
  [ $s == 'bootscripts/lfs/sysconfig/udev_retry' ] && script='config-udev-retry'; 
  
  sed  -e 's/\&/\&amp\;/g' -e 's/</\&lt\;/g'   -e 's/>/\&gt\;/g' \
       -e "s/'/\&apos\;/g" -e 's/"/\&quot\;/g' -e 's/\t/    /g'  \
       $s > appendices/${script}.script 
done


udev-lfs/init-net-rules.sh
#! /bin/bash

# This script generates rules for persistent network device naming
# Data from udev-182 75-persistent-net-generator.rules
# Updated fof udev-197 (DEVICES=en*)

RULES=/etc/udev/rules.d/70-persistent-net.rules 
DEVICES=$(eval echo /sys/class/net/{en*,eth*,ath*,wlan*[0-9],msh*,ra*,sta*,ctc*,lcs*,hsi*})

function usage
{
  echo $msg
  echo "init-net-rules.sh is an LFS-specific script to initialize"
  echo "$RULES"
  exit 1
}

declare -A VENDORS_IGNORED
VENDORS_IGNORED['52:54:00:']="kvm"
VENDORS_IGNORED['00:0c:29:']="vmware"
VENDORS_IGNORED['00:50:56:']="vmware"
VENDORS_IGNORED['00:15:5d:']="hyper-v"
VENDORS_IGNORED['00:00:00:']="invalid"

declare -A VENDORS
VENDORS['02:07:01:']="Interlan, DEC, etc"
VENDORS['02:60:60:']="3com"
VENDORS['02:60:8c:']="3Com IBM PC; Imagen. etc"
VENDORS['02:a0:c9:']="intel"
VENDORS['02:aa:3c:']="Olivetti"
VENDORS['02:cf:1f:']="Masscomp, Silicon Graphics, etc"
VENDORS['02:e0:3b:']="Gigabit"
VENDORS['02:e6:d3:']="BTI"
VENDORS['52:54:00:']="Realtek"
VENDORS['52:54:4c:']="Novell"
VENDORS['52:54:ab:']="Realtek"
VENDORS['e2:0c:0f:']="Kingston"
VENDORS['00:16:3e:']="Xensource"

function ignore_if
{
   if [[ "${VENDORS_IGNORED[$VENDOR]}" != "" ]]; then return 0; fi
   if [[ "${VENDORS[$VENDOR]}"         != "" ]]; then return 1; fi

   byte2=$(echo $VENDOR | cut -c2)
   if echo $byte2 | grep -q "[2367abef]"; then return 0; fi

   return 1  # Default is to not ignore
}

function comment
{
  # Not implemented
  # SUBSYSTEMS=="pci"
  # export COMMENT="PCI device $attr{vendor}:$attr{device} ($driver)"

  # SUBSYSTEMS=="usb", ATTRS{idVendor}=="?*"
  # export COMMENT="USB device 0x$attr{idVendor}:0x$attr{idProduct} ($driver)"

  # SUBSYSTEMS=="pcmcia",
  # export COMMENT="PCMCIA device $attr{card_id}:$attr{manf_id} ($driver)"

  # SUBSYSTEMS=="ieee1394",
  # export COMMENT="Firewire device $attr{host_id})"

  # ibmveth likes to use "locally administered" MAC addresses
  # DRIVERS=="ibmveth",
  # export COMMENT="ibmveth ($id)"

  # S/390 uses id matches only, do not use MAC address match
  # SUBSYSTEMS=="ccwgroup", 
  # export COMMENT="S/390 $driver device at $id", 
  # export MATCHID="$id" 
  # export MATCHDRV="$driver"
  # export MATCHADDR=""

  # Default
  driver=$(basename $(readlink -f $NIC/device/driver/module))
  export COMMENT="net device ${driver}"
}

if ! mountpoint -q /sys; then
  msg="/sys must be mounted"
  usage
fi

if ! mountpoint -q /proc; then
  msg="/proc must be mounted"
  usage
fi

if [ -e $RULES ]; then
  msg="The rules file already exists"
  usage
fi

# Ignore Xen virtual interfaces
if [ -e /proc/xen ]; then
  msg="The rules file should not be created in the Xen environment"
  usage
fi 

# Variables used to communicate with write_net_rules:
#   INTERFACE             simple interface name         
#   MATCHADDR             MAC address used for the match
#   MATCHID               bus_id used for the match
#   MATCHDRV              driver name used for the match
#   MATCHIFTYPE           interface type match
#   COMMENT               comment to add to the generated rule
#   INTERFACE_NAME        requested name supplied by external tool
#   INTERFACE_NEW         new interface name returned by rule writer

for NIC in $DEVICES; do
     IF=${NIC##*/}
     if echo $NIC | grep -q '*' ; then continue; fi

     export INTERFACE=${NIC##*/}            # Simple interface name
     export MATCHADDR="$(cat $NIC/address)" # Read MAC address

     VENDOR=$(echo $MATCHADDR | cut -c-9)
     if ignore_if; then continue; fi

     export MATCHDEVID="$(cat $NIC/dev_id)"
     export MATCHIFTYPE="$(cat $NIC/type)"  # Read interface type
     comment

     /lib/udev/write_net_rules 
done



lfs-latest.php
#! /usr/bin/php
<?php
$dirs = array();
$vers = array();

date_default_timezone_set( "GMT" );
$date = date( "Y-m-d H:i:s" );

// Special cases
$exceptions = array();
//$exceptions[ 'gmp' ] = "UPDIR=/.*(gmp-\d[\d\.-]*\d).*/:DOWNDIR=";

$regex = array();
//$regex[ 'bzip2'    ] = "/^.*current version is ([\d\.]+).*$/";
$regex[ 'check'    ] = "/^.*Check (\d[\d\.]+\d).*$/";
$regex[ 'intltool' ] = "/^.*Latest version is (\d[\d\.]+\d).*$/";
$regex[ 'less'     ] = "/^.*current released version is less-(\d+).*$/";
$regex[ 'mpfr'     ] = "/^mpfr-([\d\.]+)\.tar.*$/";
$regex[ 'systemd'  ] = "/^.*v([\d]+).*$/";
//$regex[ 'sysvinit' ] = "/^.*sysvinit-([\d\.]+)dsf\.tar.*$/";
$regex[ 'tzdata'   ] = "/^.*tzdata([\d]+[a-z]).*$/";
$regex[ 'xz'       ] = "/^.*xz-([\d\.]*\d).*$/";
$regex[ 'zlib'     ] = "/^.*zlib ([\d\.]*\d).*$/";

function find_max( $lines, $regex_match, $regex_replace )
{
  $a = array();
  if ( ! is_array( $lines ) ) return -1;

  foreach ( $lines as $line )
  {
     if ( ! preg_match( $regex_match, $line ) ) continue; 

     // Isolate the version and put in an array
     $slice = preg_replace( $regex_replace, "$1", $line );
     if ( $slice == $line ) continue; 

     array_push( $a, $slice );     
  }

  // SORT_NATURAL requires php-5.4.0 or later
  rsort( $a, SORT_NATURAL );  // Max version is at the top
  return ( isset( $a[0] ) ) ? $a[0] : -2;
}

function find_even_max( $lines, $regex_match, $regex_replace )
{
  $a = array();
  foreach ( $lines as $line )
  {
     if ( ! preg_match( $regex_match, $line ) ) continue;

     // Isolate the version and put in an array
     $slice = preg_replace( $regex_replace, "$1", $line );

     if ( "x$slice" == "x$line" ) continue;

     // Skip odd numbered minor versions and minors > 80
     list( $major, $minor, $rest ) = explode( ".", $slice . ".0" );
     if ( $minor % 2 == 1  ) continue;
     if ( $minor     >  80 ) continue;
     array_push( $a, $slice );
  }

  rsort( $a, SORT_NATURAL );  // Max version is at the top
  return ( isset( $a[0] ) ) ? $a[0] : -2;
}

function http_get_file( $url )
{
  if ( ! preg_match( "/sourceforge/", $url ) )
  {
    exec( "curl --location --silent --max-time 30 $url", $dir );

    $s   = implode( "\n", $dir );
    $dir = strip_tags( $s );
    return explode( "\n", $dir );
  }
  else
  {
    exec( "links -dump $url 2>/dev/null", $lines );
    return $lines;
  }
}

function max_parent( $dirpath, $prefix )
{
  // First, remove a directory
  $dirpath  = rtrim  ( $dirpath, "/" );    // Trim any trailing slash
  $position = strrpos( $dirpath, "/" );
  $dirpath  = substr ( $dirpath, 0, $position );

  $lines = http_get_file( $dirpath );

  $regex_match   = "#${prefix}[\d\.]+/#";
  $regex_replace = "#^.*(${prefix}[\d\.]+)/.*$#";
  $max           = find_max( $lines, $regex_match, $regex_replace );

  return "$dirpath/$max"; 
}

function get_packages( $package, $dirpath )
{
  global $exceptions;
  global $regex;

//if ( $package != "expat" ) return 0; // Debug

if ( $package == "check"      ) $dirpath = "https://github.com/libcheck/check/releases";
if ( $package == "e2fsprogs"  ) $dirpath = "http://sourceforge.net/projects/e2fsprogs/files/e2fsprogs";
if ( $package == "expat"      ) $dirpath = "http://sourceforge.net/projects/expat/files";
if ( $package == "elfutils"   ) $dirpath = "https://sourceware.org/ftp/elfutils";  
if ( $package == "expect"     ) $dirpath = "http://sourceforge.net/projects/expect/files";  
if ( $package == "file"       ) $dirpath = "https://github.com/file/file/releases";
if ( $package == "flex"       ) $dirpath = "https://github.com/westes/flex/releases";
if ( $package == "gcc"        ) $dirpath = max_parent( $dirpath, "gcc-" );
if ( $package == "intltool"   ) $dirpath = "https://launchpad.net/intltool/trunk";
if ( $package == "meson"      ) $dirpath = "https://github.com/mesonbuild/meson/releases";
if ( $package == "mpc"        ) $dirpath = "https://ftp.gnu.org/gnu/mpc";
if ( $package == "mpfr"       ) $dirpath = "http://mpfr.loria.fr/mpfr-current";
if ( $package == "ninja"      ) $dirpath = "https://github.com/ninja-build/ninja/releases";
if ( $package == "procps-ng"  ) $dirpath = "http://sourceforge.net/projects/procps-ng/files";
if ( $package == "psmisc"     ) $dirpath = "http://sourceforge.net/projects/$package/files";
if ( $package == "shadow"     ) $dirpath = "https://github.com/shadow-maint/shadow/releases";
if ( $package == "systemd"    ) $dirpath = "https://github.com/systemd/systemd/releases";
if ( $package == "tcl"        ) $dirpath = "http://sourceforge.net/projects/tcl/files";
if ( $package == "util-linux" ) $dirpath = max_parent( $dirpath, "v." );
if ( $package == "vim"        ) $dirpath = "ftp://ftp.vim.org/pub/vim/unix";

  // Check for ftp
  if ( preg_match( "/^ftp/", $dirpath ) ) 
  { 
    $dirpath  = substr( $dirpath, 6 );           // Remove ftp://
    $dirpath  = rtrim ( $dirpath, "/" );         // Trim any trailing slash
    $position = strpos( $dirpath, "/" );         // Divide at first slash
    $server   = substr( $dirpath, 0, $position );
    $path     = substr( $dirpath, $position );

    $conn = ftp_connect( $server );
    ftp_login( $conn, "anonymous", "" ); 

    // See if we need special handling
    if ( isset( $exceptions[ $package ] ) )
    {
       $specials = explode( ":", $exceptions[ $package ] );

       foreach ( $specials as $i )
       {
          list( $op, $regexp ) = explode( "=", $i );

          switch ($op)
          {
            case "UPDIR":
              // Remove last dir from $path
              $position = strrpos( $path, "/" );
              $path = substr( $path, 0, $position );

              // Get dir listing
              $lines = ftp_rawlist ($conn, $path);              
              $max   = find_max( $lines, $regexp, $regexp );
              break;

            case "DOWNDIR":
              // Append found directory
              $path .= "/$max";
              break;

            default:
              echo "Error in specials array for $package\n";
              return -5;
              break;
          }
       }
    }

    $lines = ftp_rawlist ($conn, $path);
    ftp_close( $conn );
  }
  else // http
  {
     // Customize http directories as needed
     if ( $package == "tzdata" )
     {
        // Remove two directories
        $dirpath  = rtrim  ( $dirpath, "/" );    // Trim any trailing slash
        $position = strrpos( $dirpath, "/" );
        $dirpath  = substr ( $dirpath, 0, $position );
        $position = strrpos( $dirpath, "/" );
        $dirpath  = substr ( $dirpath, 0, $position );
     }

     //if ( $package == "bzip2" ) 
     //{
     //   // Remove one directory
     //   $dirpath  = rtrim  ( $dirpath, "/" );    // Trim any trailing slash
     //   $position = strrpos( $dirpath, "/" );
     //   $dirpath  = substr ( $dirpath, 0, $position );
     //}

     $lines = http_get_file( $dirpath );
     if ( ! is_array( $lines ) ) return -6;
  } // End fetch
//print_r($lines);
  if ( isset( $regex[ $package ] ) )
  {
     // Custom search for latest package name
     foreach ( $lines as $l )
     {
        $ver = preg_replace( $regex[ $package ], "$1", $l );
        if ( $ver == $l ) continue;
        return $ver;  // Return first match of regex
     }

     return -7;  // This is an error
  }

  if ( $package == "perl" )  // Custom for perl
  {
     $tmp = array();

     foreach ( $lines as $l )
     {
        if ( preg_match( "/sperl/", $l ) ) continue; // Don't want this
        $ver = preg_replace( "/^.*perl-([\d\.]+\d)\.tar.*$/", "$1", $l );
        if ( $ver == $l ) continue;
        list( $s1, $s2, $rest ) = explode( ".", $ver );
        if ( $s2 % 2 == 1 ) continue; // Remove odd minor versions
        array_push( $tmp, $l );
     }

     $lines = $tmp;
  }

  if ( $package == "attr" ||  
       $package == "acl"  )
  {
     return find_max( $lines, "/$package/", "/^.*$package-([\d\.-]*\d).tar.*$/" );
  }

  if ( $package == "e2fsprogs" )
     return find_max( $lines, "/v\d/", "/^.*v(\d[\d\.]+\d).*$/" );

  if ( $package == "expect" )
     return find_max( $lines, "/expect/", "/^.*expect(\d[\d\.]+\d).tar.*$/" );

  if ( $package == "elfutils" )
     return find_max( $lines, "/^\d/", "/^(\d[\d\.]+\d)\/.*$/" );

  if ( $package == "XML-Parser" )
  {
     $max = find_max( $lines, "/$package/", "/^.*$package-([\d\._]*\d).tar.*$/" );
     # 2.44_01 is a developer release
     if ( $max == "2.44_01" ) { return "2.44"; }
     return $max;
  }

  if ( $package == "tcl" )
     return find_max( $lines, "/tcl/", "/^.*tcl(\d[\d\.]*\d)-src.*$/" );

  if ( $package == "ninja" )
     return find_max( $lines, "/v\d/", "/^.*v(\d[\d\.]*\d).*$/" );

  if ( $package == "gmp" )
     return find_max( $lines, "/$package/", "/^.*$package-([\d\._]*\d[a-z]?).tar.*$/" );

  if ( $package == "dbus" )
     return find_even_max( $lines, "/$package/", "/^.*$package-([\d\.]+).tar.*$/" );

  if ( $package == "file" )
  {
     $max = find_max( $lines, "/FILE5/", "/^.*FILE(5_\d+)*$/" );
     return str_replace( "_", ".", $max );
  }

  if ( $package == "grub" )
     return find_max( $lines, "/grub/", "/^.*grub-(\d\..*).tar.xz.*$/" );

  if ( $package == "openssl" )
     return find_max( $lines, "/openssl/", "/^.*openssl-([\d\.p]*\d.?).tar.*$/" );

  // Most packages are in the form $package-n.n.n
  // Occasionally there are dashes (e.g. 201-1)
  return find_max( $lines, "/$package/", "/^.*$package-([\d\.-]*\d)\.tar.*$/" );
}

function get_current()
{
   global $dirs;
   global $vers;

   // Fetech from svn and get wget-list
   $current = array();
   $lfssvn = "svn://svn.linuxfromscratch.org/LFS/trunk";

   $tmpdir = exec( "mktemp -d /tmp/lfscheck.XXXXXX" );
   $cdir   = getcwd();
   chdir( $tmpdir );
   exec ( "svn --quiet export $lfssvn LFS" );
   chdir( $cdir );

   $PAGE       = "$tmpdir/LFS/BOOK/chapter03/chapter03.xml";
   $STYLESHEET = "$tmpdir/LFS/BOOK/stylesheets/wget-list.xsl";

   exec( "xsltproc --xinclude --nonet $STYLESHEET $PAGE", $current );
   exec( "rm -rf $tmpdir" );

   foreach ( $current as $line )
   {
      $file = basename( $line ) . "\n";
      if ( preg_match( "/patch$/", $file ) ) { continue; } // Skip patches

      $file = preg_replace( "/bz2/", '', $file ); // The 2 confusses the regex

      $file        = rtrim( $file );
      $pkg_pattern = "/(\D*).*/";
      //$pattern     = "/\D*(\d.*\d)\D*/";
      $pattern     = "/\D*(\d.*\d)\D*/";

      if ( preg_match( "/e2fsprogs/", $file ) )
      {
        $pattern = "/e2\D*(\d.*\d)\D*/";
        $pkg_pattern = "/(e2\D*).*/";
      }

      else if ( preg_match( "/tzdata/", $file ) )
      {
        $pattern = "/\D*(\d.*[a-z])\.tar\D*/";
      }

      else if ( preg_match( "/openssl/", $file ) )
      {
        $pattern = "/\D*(\d.*\d.*).tar.*$/";
      }

      else if ( preg_match( "/gmp/", $file ) )
      {
        $pattern = "/\D*(\d.*[a-z]*)\.tar\D*/";
      }

      else if ( preg_match( "/systemd-man-pages/", $file ) ) continue; 
      else if ( preg_match( "/python/"         , $file ) ) continue; 

      $version = preg_replace( $pattern, "$1", $file );   // Isolate version
      $version = preg_replace( "/^\d-/", "", $version );  // Remove leading #-

      // Touch up package names
      $pkg_name = preg_replace( $pkg_pattern, "$1", $file );
      $pkg_name = trim( $pkg_name, "-" );

      if ( preg_match( "/bzip|iproute/", $pkg_name ) ) { $pkg_name .= "2"; }
      if ( preg_match( "/^m$/"         , $pkg_name ) ) { $pkg_name .= "4"; }
      if ( preg_match( "/shadow/"      , $pkg_name ) ) { $pkg_name  = "shadow"; }

      $dirs[ $pkg_name ] = dirname( $line );
      $vers[ $pkg_name ] = $version;
   }
}

function mail_to_lfs()
{
   global $date;
   global $vers;
   global $dirs;

   //$to      = "bruce.dubbs@gmail.com";
   $to      = "lfs-book@lists.linuxfromscratch.org";
   $from    = "bdubbs@linuxfromscratch.org";
   $subject = "LFS Package Currency Check - $date GMT";
   $headers = "From: bdubbs@anduin.linuxfromscratch.org";

   $message = "Package         LFS      Upstream  Flag\n\n";

   foreach ( $dirs as $pkg => $dir )
   {
      //if ( $pkg != "gmp" ) continue;  //debug
      $v = get_packages( $pkg, $dir );

      $flag = ( $vers[ $pkg ] != $v ) ? "*" : "";

      // Pad for output
      $pad = "                ";
      $p   = substr( $pkg          . $pad, 0, 15 );
      $l   = substr( $vers[ $pkg ] . $pad, 0, 10 );
      $c   = substr( $v            . $pad, 0, 10 );

      $message .= "$p $l $c $flag\n";
   }

   exec ( "echo '$message' | mailx -r $from -s '$subject' $to" );
   //echo $message;
}

function html()
{

   global $date;
   global $vers;
   global $dirs;

   echo "<!DOCTYPE html PUBLIC '-//W3C//DTD XHTML 1.0 Strict//EN'
                      'http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd'>
<html xmlns='http://www.w3.org/1999/xhtml' xml:lang='en' lang='en'>
<head>
<title>LFS Package Currency Check - $date</title>
<style type='text/css'>
h1, h2 {
   text-align      : center;
}

table {
   border-width    : 1px;
   border-spacing  : 0px;
   border-style    : outset;
   border-color    : gray;
   border-collapse : separate;
   background-color: white;
   margin          : 0px auto;
}

table th {
   border-width    : 1px;
   padding         : 2px;
   border-style    : inset;
   border-color    : gray;
   background-color: white;
}

table td {
   border-width    : 1px;
   padding         : 2px;
   border-style    : inset;
   border-color    : gray;
   background-color: white;
}
</style>

</head>
<body>
<h1>LFS Package Currency Check</h1>
<h2>As of $date GMT</h1>

<table>
<tr><th>LFS Package</th> <th>LFS Version</th> <th>Latest</th> <th>Flag</th></tr>\n";

   // Get the latest version of each package
   foreach ( $dirs as $pkg => $dir )
   {
      $v    = get_packages( $pkg, $dir );
      $flag = ( $vers[ $pkg ] != $v ) ? "*" : "";
      echo "<tr><td>$pkg</td> <td>${vers[ $pkg ]}</td> <td>$v</td> <td>$flag</td></tr>\n";
   }

   echo "</table>
</body>
</html>\n";

}

get_current();  // Get what is in the book
mail_to_lfs();
//html();  // Write html output
?>
